File: springboot-jwt/src/main/java/com/grepp/spring/app/model/member/MemberService.java
package com.grepp.spring.app.model.member;

import com.grepp.spring.app.model.auth.code.Role;
import com.grepp.spring.app.model.member.dto.MemberDto;
import com.grepp.spring.app.model.member.entity.Member;
import com.grepp.spring.app.model.member.entity.MemberInfo;
import com.grepp.spring.infra.error.exceptions.CommonException;
import com.grepp.spring.infra.mail.MailTemplate;
import com.grepp.spring.infra.mail.SmtpDto;
import com.grepp.spring.infra.response.ResponseCode;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Slf4j
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class MemberService{

    private final PasswordEncoder passwordEncoder;
    private final MemberRepository memberRepository;
    private final ModelMapper mapper;
    private final MailTemplate mailTemplate;

    @Value("${app.domain}")
    private String domain;

    @Transactional
    public void signup(MemberDto dto, Role role) {
        if(memberRepository.existsById(dto.getUserId()))
            throw new CommonException(ResponseCode.BAD_REQUEST);

        Member member = mapper.map(dto, Member.class);

        String encodedPassword = passwordEncoder.encode(dto.getPassword());
        member.setPassword(encodedPassword);
        member.setRole(role);

        MemberInfo memberInfo = new MemberInfo();
        memberInfo.setUserId(dto.getUserId());
        member.setInfo(memberInfo);
        memberRepository.save(member);
    }

    public Boolean isDuplicatedId(String id) {
        return memberRepository.existsById(id);
    }

    public MemberDto findById(String userId) {
        Member member = memberRepository.findById(userId)
            .orElseThrow(() -> new CommonException(ResponseCode.BAD_REQUEST));
        return mapper.map(member, MemberDto.class);
    }

    public void sendVerificationMail(String token, MemberDto dto) {
        if(memberRepository.existsById(dto.getUserId()))
            throw new CommonException(ResponseCode.BAD_REQUEST);

        SmtpDto smtpDto = new SmtpDto();
        smtpDto.setTo(dto.getEmail());
        smtpDto.setTemplatePath("/mail/signup-verification");
        smtpDto.setSubject("회원가입을 환영합니다!");
        smtpDto.setProperties(Map.of("domain", domain, "token", token));
        mailTemplate.send(smtpDto);
    }
}

File: user-service/src/main/kotlin/com/grepp/app/model/member/MemberService.kt
package com.grepp.spring.app.model.member

import com.grepp.app.model.member.code.Role
import com.grepp.infra.response.ResponseCode
import com.grepp.spring.app.controller.api.member.payload.MemberDetailResponse
import com.grepp.spring.app.controller.api.member.payload.SignupRequest
import com.grepp.spring.app.model.member.entity.Member
import com.grepp.spring.app.model.member.entity.MemberInfo
import com.grepp.spring.infra.error.exceptions.CommonException
import org.modelmapper.ModelMapper
import org.slf4j.LoggerFactory
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class MemberService(
    private val memberRepository: MemberRepository,
    private val passwordEncoder: PasswordEncoder,
    private val mapper: ModelMapper
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Transactional
    fun signup(request:SignupRequest) {
        if(memberRepository.existsById(request.userId))
            throw CommonException(ResponseCode.BAD_REQUEST)

        val encodedPassword = passwordEncoder.encode(request.password)

        val member = Member(request.userId, encodedPassword,
            request.email, Role.ROLE_USER, request.tel)

        val memberInfo = MemberInfo(request.userId)
        member.info = memberInfo
        memberRepository.save(member)
    }

    fun checkId(id: String): Boolean {
        return memberRepository.existsById(id)
    }

    fun findById(id: String): MemberDetailResponse {
        val member = memberRepository.findById(id)
        return mapper.map(member, MemberDetailResponse::class.java)
    }


}